### PR TITLE
Fix for screenshots with function scoped browsers

### PR DESCRIPTION
**Pull request by pelme**
See issue #74

---

This change is [Reviewable](https://reviewable.io/reviews/pytest-dev/pytest-splinter/75)

### CODE REVIEW

**Comment by pelme** [comment]
I'm not sure why https://travis-ci.org/pytest-dev/pytest-splinter/jobs/165209264#L331-L346 happens, it only seems to happen when multiple tests run.

I can run `test_browser_screenshot_xdist` locally without failures with and without xdist installed:

```
‹pytest-splinter› ‹function-scoped-screenshot› ~/code/pytest-splinter → py.test -k test_browser_screenshot_xdist
=============================================================== test session starts ===============================================================
platform darwin -- Python 3.5.0, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- /Users/andreas/.virtualenvs/pytest-splinter/bin/python
cachedir: .cache
rootdir: /Users/andreas/code/pytest-splinter, inifile: tox.ini
plugins: pylama-7.1.0, cov-2.3.1, localserver-0.3.6, xdist-1.15.0, splinter-1.7.6
collected 34 items

tests/mocked_browser/test_screenshot.py::test_browser_screenshot_xdist PASSED

=============================================================== 33 tests deselected ===============================================================
===================================================== 1 passed, 33 deselected in 2.92 seconds =====================================================
```

```
‹pytest-splinter› ‹function-scoped-screenshot› ~/code/pytest-splinter → pip uninstall pytest-xdist
Uninstalling pytest-xdist:
  ....
Proceed (y/n)? y
  Successfully uninstalled pytest-xdist
```

```
‹pytest-splinter› ‹function-scoped-screenshot› ~/code/pytest-splinter → py.test -k test_browser_screenshot_xdist
=============================================================== test session starts ===============================================================
platform darwin -- Python 3.5.0, pytest-3.0.3, py-1.4.31, pluggy-0.4.0 -- /Users/andreas/.virtualenvs/pytest-splinter/bin/python
cachedir: .cache
rootdir: /Users/andreas/code/pytest-splinter, inifile: tox.ini
plugins: pylama-7.1.0, cov-2.3.1, localserver-0.3.6, splinter-1.7.6
collected 34 items

tests/mocked_browser/test_screenshot.py::test_browser_screenshot_xdist SKIPPED

=============================================================== 33 tests deselected ===============================================================
==================================================== 1 skipped, 33 deselected in 0.20 seconds =====================================================
```

**Comment by bubenkoff** [comment]
tests are failing, please fix

**Comment by pelme** [comment]
Can you give the diff/failure a quick look, maybe you could figure out why? I will try to dig deeper but it was not straightforward to reproduce.

**Comment by bubenkoff** [comment]
tests/test_plugin.py:198:0: C0301 Line too long (123/120) [pylint] :)

**Comment by pelme** [comment]
@bubenkoff the line length is now fixed

I am still puzzled about the failure which only happens when I run the full test suite and not just the xdist test. I think the take_screenshot finalizer sometimes ends up as a session finalizer (which is not desirable). Do you have any ideas on where to put the screenshot-taking code?

File: pytest_splinter/plugin.py [hunk 1]
```diff
@@ -398,6 +452,12 @@ def prepare_browser(request, parent):
                 request.addfinalizer(browser.quit)
         elif not browser:
             browser = browser_pool[browser_key] = get_browser(splinter_webdriver)
+        request.addfinalizer(functools.partial(
```
Comment (bubenkoff) [inline pytest_splinter/plugin.py]: so for session scoped browsers, this fails due to the `pytest_runtest_makereport` hook doing nothing useful (as it's session's request and not node's request)

**Comment by pelme** [comment]
@bubenkoff please have another look. I've moved things around a bit to call the screenshot capture at the correct location for both session and function scoped browsers. I can cleanup the commit history a bit if you think this approach is good.

**Comment by bubenkoff** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pytest_splinter/plugin.py, line 339 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_3dahj1lMEhl2p1XB:-KV_3dahj1lMEhl2p1XC:b-pc6449) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L339)):*
> ```Python
> 
> 
> def _take_screenshot(
> ```

document the function

---

*[pytest_splinter/plugin.py, line 349 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_4FCR4ex-EP00Sxof:-KV_4FCR4ex-EP00Sxog:b-ynhenr) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L349)):*
> ```Python
> ):
> 
>     if splinter_make_screenshot_on_failure and request.node.splinter_failure:
> ```

move the condition to the fixture
this function then should be clearer what it does - makes screenshots
and the decision whether to make a screenshot or not stays outside

---

*[pytest_splinter/plugin.py, line 422 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_4TdxtFq4do1ixx2T:-KV_4TdxtFq4do1ixx2U:b-euttlx) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L422)):*
> ```Python
> 
>     for name, value in fixture_values.items():
>         if hasattr(value, '__splinter_browser__'):
> ```

as commented, here should be the decision whether to make a screenshot

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75)*

**Comment by pelme** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pytest_splinter/plugin.py, line 455 at r2](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KTJcOKkisNxZvkZ7-aW-r2-455:-KV_OR6prWALAkfAvO0B:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/0c34083b51d2c4428a4115c1144bcf1d66ec9733/pytest_splinter/plugin.py#L455)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary>
> so for session scoped browsers, this fails due to the `pytest_runtest_makereport` hook doing nothing useful (as it's session's request and not node's request)
</details>
Done.

---

*[pytest_splinter/plugin.py, line 339 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_3dahj1lMEhl2p1XB:-KV_OIHt7ljaWrmBwKci:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L339)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary>
> document the function
</details>
Done.

---

*[pytest_splinter/plugin.py, line 349 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_4FCR4ex-EP00Sxof:-KV_ONgArmgnNZdyByhu:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L349)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary>
> move the condition to the fixture
> this function then should be clearer what it does - makes screenshots
> and the decision whether to make a screenshot or not stays outside
</details>
Done.

---

*[pytest_splinter/plugin.py, line 422 at r4](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75#-KV_4TdxtFq4do1ixx2T:-KV_OOPn1DB3OK2qmMdx:b-896fix) ([raw file](https://github.com/pytest-dev/pytest-splinter/blob/21de5e4ff1c2186ecb7245b5be14d682171ec552/pytest_splinter/plugin.py#L422)):*
<details><summary><i>Previously, bubenkoff (Anatoly Bubenkov) wrote…</i></summary>
> as commented, here should be the decision whether to make a screenshot
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75)*

**Comment by bubenkoff** [comment]
Reviewed 1 of 2 files at r4, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pytest-dev/pytest-splinter/75)*

**Comment by bubenkoff** [comment]
thanks!
please release it (and add changelog entry)

**Pull request merged**